### DISCUSSION TITLE
Visualization and controll enhancements

### DISCUSSION THREAD

**Discussion by calabr** [terjeio/ioSender#160] [Ideas]
Trying to test ioSender I found lack of some information and and controlls.
 It's good to be:
 Drow machine limit outline on 3D view.
 Drow tool position on screen when programm file not loaded.
 Drow crosshair instead of tool cone when tool picture become to small - if we zoom-out screen to see large area tool cone become to small and it's impossible to lacate it on the screen. With zoom to see 200mm tool is completely disapeared.
Show MPos coordinates on DRO frame.
Show current configured G1 Feed rate and spindle speed (not only when moving and spindle is running)
Show modal command state in DRO
Make side tabs, like Jog, macros etc. datachable and awailable to place on screen permanently.
Make optiomal comfirmation for macros buttons.
Remove rare ose tabs, like app settings, GRBL settings, driver settings etc. from main screen and hide it to menu.

**Comment by phil-barrett** [comment]
Some of these have come up in the past.  

> Drow machine limit outline on 3D view.

useful. though limits gives a good sense of it.

> Drow tool position on screen when programm file not loaded.

Not sure how helpful this is.  If there was a camera mode, it would be useful.

> Drow crosshair instead of tool cone when tool picture become to small - if we zoom-out screen to see large area tool cone become to 
small and it's impossible to lacate it on the screen. With zoom to see 200mm tool is completely disapeared.

Better to change tool cone size.  If looking at an angle, the cursor becomes hard to understand.

> Show current configured G1 Feed rate and spindle speed (not only when moving and spindle is running)

Is this possible?  It seems like the controller reports 0 as the feed rate when not moving.  Perhaps ioSender could track the current feed rate by watching for an Fxxx but it would have to track actual execution of it.  

> Make side tabs, like Jog, macros etc. datachable and awailable to place on screen permanently.

I'd like to see the jog panel be detachable/permanent.  I suspect there is a structure issue in ioSender to make it work that way though.

**Reply by terjeio** [reply]
>> Show current configured G1 Feed rate and spindle speed (not only when moving and spindle is running)

> Is this possible?

With grblHAL you can enable _Parser state_ in _$10 Status report options_ to get both plus the active state in most other modal groups.
This info is _pushed_ by the controller on changes. 

![image](https://user-images.githubusercontent.com/20260062/142504432-9a53a38e-2729-41ec-ac5d-9033caf564d2.png)

> Perhaps ioSender could track the current feed rate by watching for an Fxxx but it would have to track actual execution of it.

I do not want to track it from the gcode as it is sent, Grbl users should either upgrade to grblHAL or get Grbl updated to push the parser state on changes. One reason for this is that there might be a MPG changing these values, having the controller pushing them ensures the sender is kept up to date.

**Reply by phil-barrett** [reply]
I see.  And it shows it even after program end.

**Reply by calabr** [reply]
> This info is pushed by the controller on changes.

Why not to ask controller periodically (for GRBL), like most of senders do?

>> Drow tool position on screen when programm file not loaded.

>@phil-barrett Not sure how helpful this is. If there was a camera mode, it would be useful.

 Even if programm and workpeese not loaed, tool and spindle is there and can be moved - so I should see it.
 Currently my machime stands on the back on me. I cannot see screen and tool in the same time. If I see tool position and working area border on screen, I can position it as I need without turn-over.
 Also I found very useful GoTo mouse position function ( mainly I use bCNC program)

**Comment by terjeio** [comment]
> Show MPos coordinates on DRO frame.

 For now you can enable text overlay in the 3D View - this contains both MPos and WPos:
![image](https://user-images.githubusercontent.com/20260062/142505980-aecb2b57-dd9e-46df-a5cc-e65fbb86321f.png)

> Show modal command state in DRO

Is this what is shown in the [reply above](https://github.com/terjeio/ioSender/discussions/160#discussioncomment-1667450)?

Updating the 3D View is in my plans, improved grid and scaling the tool is part of that. I will consider some of your other suggestions, but I would like to have a rationale for moving the settings tabs to menus.

**Reply by calabr** [reply]
> For now you can enable text overlay in the 3D View - this contains both MPos and WPos:

 I have done it, but found it hard to read. 
> Is this what is shown in the reply above?

 Yes thet is it, but for me it's looks more usfull to group most of machine parameters in the same pannel.

> I will consider some of your other suggestions, but I would like to have a rationale for moving the settings tabs to menus.

  Im my vision, main screan should only contain elements, needed for everyday machine use. programm and machine setup will be used short time during configuratin and than not needed for long time.
 Also DRO and main controlls should alwais be visible in the same form, size and position, especialy when we want to to do something with machine position (probbing, jogging, offsets etc.)
